src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs

```diff
@@ -181,69 +144,130 @@ public void Close()
 
             allRunsCompleted = testRunCompleteArgs.IsCanceled || _abortRequested
                 ? _runCompletedClients == _runStartedClients
-                : _runCompletedClients == _availableTestSources;
+                : _runCompletedClients == _availableWorkloads;
 
             EqtTrace.Verbose("ParallelProxyExecutionManager: HandlePartialRunComplete: Total completed clients = {0}, Run complete = {1}, Run canceled: {2}.", _runCompletedClients, allRunsCompleted, testRunCompleteArgs.IsCanceled);
         }
 
-        // verify that all executors are done with the execution and there are no more sources/testcases to execute
         if (allRunsCompleted)
         {
-            // Reset enumerators
-            _sourceEnumerator = null;
-            _testCaseListEnumerator = null;
-
-            _currentRunDataAggregator = null;
-            _currentRunEventsHandler = null;
-
-            // Dispose concurrent executors
-            // Do not do the cleanup task in the current thread as we will unnecessarily add to execution time
-            UpdateParallelLevel(0);
-
+            _parallelOperationManager.StopAllManagers();
             return true;
         }
 
-
-        EqtTrace.Verbose("ParallelProxyExecutionManager: HandlePartialRunComplete: Replace execution manager. Shared: {0}, Aborted: {1}.", SharedHosts, testRunCompleteArgs.IsAborted);
-
-        RemoveManager(proxyExecutionManager);
-        proxyExecutionManager = CreateNewConcurrentManager();
-        var parallelEventsHandler = GetEventsHandler(proxyExecutionManager);
-        AddManager(proxyExecutionManager, parallelEventsHandler);
-
         // If cancel is triggered for any one run or abort is requested by test platform, there is no reason to fetch next source
-        // and queue another test run
+        // and queue another test run.
         if (!testRunCompleteArgs.IsCanceled && !_abortRequested)
         {
-            StartTestRunOnConcurrentManager(proxyExecutionManager);
+            _parallelOperationManager.RunNextWork(proxyExecutionManager);
+            //var hadMoreWork = _parallelOperationManager.RunNextWork(proxyExecutionManager);
```

Comment:
Shall we clean this?

Reply:
I did. Added comment with why returning true is bad, and kept the original code in the comment to show what I meant.